src/Microsoft.AspNetCore.ResponseCaching/Streams/WriteOnlyShardStream.cs

```diff
@@ -0,0 +1,215 @@
+            }
+            return _shards;
+        }
+
+        public override bool CanRead => false;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => !_closed;
+
+        public override long Length => _length;
+
+        public override long Position
+        {
+            get
+            {
+                return _length;
+            }
+            set
+            {
+                throw new NotSupportedException("The stream does not support seeking.");
+            }
+        }
+
+        private void DisposeMemoryStream()
+        {
+            // Clean up the memory stream
+            _bufferStream.SetLength(0);
+            _bufferStream.Capacity = 0;
+            _bufferStream.Dispose();
+        }
+
+        private void FinalizeShards()
+        {
+            // Append any remaining shards
+            if (_bufferStream.Length > 0)
+            {
+                // Add the last shard
+                _shards.Add(_bufferStream.ToArray());
+            }
+
+            DisposeMemoryStream();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                if (disposing)
+                {
+                    _shards.Clear();
+                    DisposeMemoryStream();
+                }
+
+                _disposed = true;
+                _closed = true;
+            }
+            finally
+            {
+                base.Dispose(disposing);
+            }
+        }
+
+        public override void Flush()
+        {
+            if (!CanWrite)
+            {
+                throw new ObjectDisposedException("The stream has been closed for writing.");
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException("The stream does not support reading.");
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException("The stream does not support seeking.");
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException("The stream does not support seeking.");
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset), offset, "Non-negative number required.");
+            }
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), count, "Non-negative number required.");
+            }
+            if (buffer.Length < offset + count)
```

Comment:
Always use `count > length - offset` to avoid integer overflows in `offset + count`.